infra/Pos/Communication/Relay/Logic.hs

```diff
@@ -310,38 +319,107 @@ invDataListener enqueue InvReqDataParams{..} = listenerConv $ \__ourVerInfo node
             inv' <- recvLimited conv
             whenJust inv' $ expectInv $ \InvMsg{..} -> do
                 useful <- handleInvDo (handleInv nodeId) imKey
-                whenJust useful $ \ne -> do
-                    send conv $ ReqMsg ne
-                    dt' <- recvLimited conv
-                    whenJust dt' $ expectData $ \DataMsg{..} -> do
-                          handleDataDo nodeId invReqMsgType enqueue contentsToKey (handleData nodeId) dmContents
-                          -- handlingLoop
-
-                          -- TODO CSL-1148 Improve relaing: support multiple data
-                          -- Need to receive Inv and Data messages simultaneously
-                          -- Maintain state of sent Reqs
-                          -- And check data we are sent is what we expect (currently not)
+                case useful of
+                    Nothing -> send conv (Left (ReqMsg Nothing))
+                    Just ne -> do
+                        send conv $ Left (ReqMsg (Just ne))
+                        dt' <- recvLimited conv
+                        whenJust dt' $ expectData $ \DataMsg{..} -> do
+                              res <- handleDataDo nodeId invReqMsgType enqueue contentsToKey (handleData nodeId) dmContents
```

Comment:
It could, if the particular data handler (`handleData`) throws an exception. In that case the conversation will close up and the other side will get a premature exit, indicating that something went awry.